source/mongodb-handshake/handshake.rst

```diff
@@ -224,6 +224,99 @@ Example::
         - "Oracle JVM EE 9.1.1"
 
 
+--------------------------
+Speculative Authentication
+--------------------------
+
+:since: 4.4
+
+The ``isMaster`` handshake supports a new argument, ``speculativeAuthenticate``,
+provided as a BSON object. Clients specifying this argument to ``isMaster`` will
+speculatively include the first command of an authentication handshake.
+This command may be provided to the server in parallel with any standard request for
+supported authentication mechanisms (i.e. ``saslSupportedMechs``). This would permit
+clients to merge their first message with their ``isMaster`` request, and
+receive the authentication reply with the ``isMaster`` reply. This argument has the
+following structure::
+
+    {
+        isMaster: 1,
+        speculativeAuthenticate: {
+            /* OPTIONAL */
+            authenticate: 1
+            /* OPTIONAL. If present, "payload" and "options" are required */
+            saslStart: 1
+            /* REQUIRED */
+            mechanism: "<string>"
+            /* REQUIRED */
+            db: "<string>"
+            payload: BinData(...)
+            options: {
+                skipEmptyExchange: true         /* REQUIRED */        
+            }
+        }
+    }
```

Comment:
I had trouble reading this because it merges the x509 and SCRAM speculativeAuthenticate documents. What do you think about splitting the description like this:

```
When the mechanism is ``MONGODB-X509``, `speculativeAuthenticate` has the following structure::
    {
        isMaster: 1,
        speculativeAuthenticate: {
            authenticate: 1,
            mechanism: "MONGODB-X509"
        }
    }

When the mechanism is ``SCRAM-SHA-1`` or ``SCRAM-SHA-256``, `speculativeAuthenticate` has the following structure::
    {
        isMaster: 1,
        speculativeAuthenticate: {
            saslStart: 1,
            mechanism: "<string>",   /* "SCRAM-SHA-1" or "SCRAM-SHA-256" */
            db: "<authentication database>",
            payload: BinData(...),
            options: {
                skipEmptyExchange: true         /* REQUIRED */
            }
        }
    }
```

Reply:
You are correct. Your suggestion reads much better. I have revised the text.

Reply:
Minor nit; `skipEmptyExchange` isn't actually required from a technical point of view.  It's highly encouraged (because why save one packet if you're just going to waste another), but omitting it will still function in a 3 round trip exchange.